README.md
# sakura
[![Build Status](https://travis-ci.org/leannmak/sakura.svg?branch=master)](https://travis-ci.org/leannmak/sakura)
[![Coverage Status](https://coveralls.io/repos/github/leannmak/sakura/badge.svg?branch=master)](https://coveralls.io/github/leannmak/sakura?branch=master)  

**sakura**, named from **sakura kinomoto**, is developed as a tool for configuration file management based on `Etcd`, `Confd` and `Ansible`.

## External Tool Required
```shell
1. Etcd (better use TLS)
2. Confd
3. MySQL (or other databases)
4. RabbitMQ
5. Minio
```

## Setup

```shell
$ git clone https://github.com/leannmak/sakura.git
$ git branch dev
$ cd sakura
$ pip install -r requirements.txt
$ vi config.py
$ python manage.py recreatedb
$ python runserver.py
$ export PYTHONOPTIMIZE=1
$ celery worker -B -A  sakura.celery --loglevel=debug --pidfile=.tmp/sakura-celery.pid -f .log/sakura-celery.log -c 2 -Q sakura
```

### Run Test

```shell
$ pip install tox
$ vi instance/test_config.py
$ tox
```

### Run Script

```shell
$ python manage.py --help
usage: manager script support for sakura

manager script support for sakura

positional arguments:
  {shell,db,recreatedb,initdb,runserver,dropdb}
    shell               Runs a Python shell inside Flask application context.
    db                  Perform database migrations
    recreatedb          Recreates database tables (same as issuing 'dropdb'
                        and 'initdb')
    initdb              Initialize database tables
    runserver           Runs the Flask development server i.e. app.run()
    dropdb              Drops database tables

optional arguments:
  -?, --help            show this help message and exit
```

### API Usage

See [realtime.md](document/realtime.md) and [history.md](document/history.md).


document/history.md
# SAKURA API Usage - History Module

Leann Mak, leannmak@139.com, (c) 2018.

## Notice
 
Remember to set ```Content-Type: application/json``` when `POST` if using `JSON`.

| Role | Name | Location | Type | Description |
| --- | --- | --- | ----- | ----------- |
| 数据格式 | Content-Type | header | application/json | POST/PUT请求数据格式为JSON |

## API List
* [task（任务执行记录）](#task)
    * [GET: retrieving tasks](#taskget)

### task
#### task.get
##### /api/&lt;api version&gt;/sakura/task
* Description: this api allows to retrieve tasks.
* Normal response code: 200
* Error response code: 400, 500
* Error message:

| Message | Meaning | Code |
| --------------- | --------------- | --- |
| Invalid Access | 请求参数/格式非法 | 400 |

* Request arguments:

| Role | Name | Location | Type | Description | Required |
| --- | ------ | --- | --- | ----------- | --- |
| 分页查询 | page | URL | integer | 访问页码 | no |
| 分页查询 | pp | URL | integer | 每页记录数，指定page后生效，默认为20 | no |
| 信息检索 | 任意非扩展字段名称 | URL | string | 所有非扩展字段可检索 | no |

* Return values:

| Role | Name | Location | Type | Description | Always in |
| --- | --- | --- | --- | ----------- | --- |
| 数据信息 | data | body | dictionary | 查询结果 | yes |
| 分页信息 | totalpage | body | integer | 总页码数，若page未生效，默认为false | yes |

* Examples:  

Request:

```http
GET /api/v1.0/sakura/task
```

Response:

```json
{
    "data": [
        {
            "ack_status": null,
            "begin_time": "Fri, 08 Jun 2018 15:43:50 GMT",
            "delta_time": 3.4856,
            "end_time": "Fri, 08 Jun 2018 15:43:53 GMT",
            "info": "{\"current\": 100, \"message\": \"Configurations files have been checked completely.\", \"total\": 100, \"data\": [{\"greeting.cfg\": {\"127.0.0.1\": {\"content_actual\": \"hello jude, my name is leann.\\nnice to meet you.\\n \\nme too.\", \"content_expected\": \"hello world!\", \"last_modify_time\": \"2018-05-22 10:24:10.637048142 +0800\", \"content\": \"f6ee91aecaa4e412dd75dacff705c9ea != fc3ff98e8c6a0d3087d515c0473f8677\", \"mode\": \"0775 != 0755\", \"owner\": \"OK\"}}}, {\"test.cfg\": {\"127.0.0.1\": {\"content\": \"OK\", \"owner\": \"OK\", \"mode\": \"OK\", \"last_modify_time\": \"2018-06-08 14:35:38.159246182 +0800\"}}}]}",
            "kwargs": "{\"files\": [{\"name\": \"greeting.cfg\", \"items\": {}, \"mode\": \"0755\", \"template\": \"hello world!\", \"owner\": {\"group\": \"apps\", \"name\": \"appa\"}, \"dir\": \"/apps/conf/test\"}, {\"name\": \"test.cfg\", \"items\": {\"what\": \"is playing basketball\", \"who\": \"jay\", \"when\": \"now\", \"whom\": \"kobe\", \"where\": \"on the playground\", \"why\": \"for fun\"}, \"mode\": \"0755\", \"template\": \"{{getv \\\"/who\\\"}} {{getv \\\"/what\\\"}} with {{getv \\\"/whom\\\"}} {{getv \\\"/where\\\"}} {{getv \\\"/when\\\"}} {{getv \\\"/why\\\"}}.\\r\\n\", \"owner\": {\"group\": \"leannmak\", \"name\": \"leannmak\"}, \"dir\": \"/apps/conf/test\"}], \"hosts\": [\"127.0.0.1\"]}",
            "name": "configuration_check",
            "state": "SUCCESS",
            "step": 0,
            "sub_task_id": null,
            "task_id": "d90c3081-3fbf-4058-9545-54d7b69e78bc"
        },
        {
            "ack_status": null,
            "begin_time": "Fri, 08 Jun 2018 15:19:07 GMT",
            "delta_time": 2.38401,
            "end_time": "Fri, 08 Jun 2018 15:19:09 GMT",
            "info": "{\"current\": 100, \"message\": \"Task <7640edbc-d6b2-4fc7-ab37-6f8a3c16bb40> have been acknowledged.\", \"total\": 100, \"data\": null}",
            "kwargs": "{\"main_task_id\": \"7640edbc-d6b2-4fc7-ab37-6f8a3c16bb40\"}",
            "name": "configuration_acknowledge",
            "state": "SUCCESS",
            "step": 0,
            "sub_task_id": null,
            "task_id": "a3ea15d7-ec83-44c6-bb35-5689f94d3887"
        },
        {
            "ack_status": "PASSED",
            "begin_time": "Fri, 08 Jun 2018 14:35:29 GMT",
            "delta_time": 8.84253,
            "end_time": "Fri, 08 Jun 2018 14:35:38 GMT",
            "info": "{\"current\": 100, \"message\": \"Configurations have been updated completely.\", \"total\": 100, \"data\": null}",
            "kwargs": "{\"files\": [{\"name\": \"test.cfg\", \"items\": {\"what\": \"is playing basketball\", \"who\": \"jay\", \"when\": \"now\", \"whom\": \"kobe\", \"where\": \"on the playground\", \"why\": \"for fun\"}, \"mode\": \"0755\", \"template\": \"{{getv \\\"/who\\\"}} {{getv \\\"/what\\\"}} with {{getv \\\"/whom\\\"}} {{getv \\\"/where\\\"}} {{getv \\\"/when\\\"}} {{getv \\\"/why\\\"}}.\\r\\n\", \"owner\": {\"group\": \"leannmak\", \"name\": \"leannmak\"}, \"dir\": \"/apps/conf/test\"}], \"use_disconf\": false, \"service_version\": \"1.0\", \"service_name\": \"test\", \"check_cmd\": \"\", \"hosts\": [\"127.0.0.1\"], \"env_name\": \"qa\", \"reload_cmd\": \"\"}",
            "name": "configuration_update",
            "state": "SUCCESS",
            "step": 4,
            "sub_task_id": "a3ea15d7-ec83-44c6-bb35-5689f94d3887",
            "task_id": "7640edbc-d6b2-4fc7-ab37-6f8a3c16bb40"
        }
    ],
    "totalpage": false
}
```


sakura/service.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
#
# Leann Mak, leannmak@139.com, (c) 2018.
# This is the service module of sakura package.
#

import os
import json
import werkzeug
import traceback
from datetime import datetime
from inspect import isclass, isfunction
from flask.ext.restful import reqparse, Resource, inputs

from sakura import app
from sakura import constant as C
from sakura.model import TaskManager
from sakura.task import SakuraTask
from sakura.util import logmsg
from sakura.input import defined_dictionary, union_dictionary, ip


""" Error
"""


class SakuraAPIError(Exception):
    __abstract__ = True

    def __init__(self, prefix, content, http_code):
        self.message = '%s: %s' % (prefix, content)
        self.code = http_code

    def __str__(self):
        return repr(self.message)


class SakuraInvalidAccessError(SakuraAPIError):
    def __init__(self, content='Bad Request Data Format'):
        prefix = 'Invalid Access'
        code = 400
        super(SakuraInvalidAccessError, self).__init__(
            prefix=prefix, content=content, http_code=code)


class SakuraConstraintConflictError(SakuraAPIError):
    def __init__(self, content):
        prefix = 'Task Constraint Conflict'
        code = 403
        super(SakuraConstraintConflictError, self).__init__(
            prefix=prefix, content=content, http_code=code)


class SakuraObjectNotFoundError(SakuraAPIError):
    def __init__(self, content):
        prefix = 'Object Not Found'
        code = 404
        super(SakuraObjectNotFoundError, self).__init__(
            prefix=prefix, content=content, http_code=code)


""" Service
"""


class TaskListAPI(Resource):
    """
        TaskList Restful API.
        For GET(Readonly).
    """
    def __init__(self):
        super(TaskListAPI, self).__init__()
        self.obj = TaskManager()
        self.parser = reqparse.RequestParser(bundle_errors=True)
        # page
        self.parser.add_argument(
            'page', type=inputs.positive,
            help='Page must be a positive integer')
        # pp: number of items per page
        self.parser.add_argument(
            'pp', type=inputs.positive,
            help='PerPage must be a positive integer', dest='per_page')
        # multi-type parameters
        setattr(self, 'params', [])
        type_dict = {'String': unicode, 'Integer': int}
        for k, v in self.obj._columns().items():
            if v.type.__class__.__name__ in type_dict:
                self.parser.add_argument(
                    k, type=type_dict[v.type.__class__.__name__])
                self.params.append(k)

    def get(self):
        """ get whole list of tasks
        """
        try:
            try:
                args = self.parser.parse_args()
            except Exception as e:
                if hasattr(e, 'data'):
                    raise SakuraInvalidAccessError(e.data['message'])
                else:
                    raise SakuraInvalidAccessError()
            pages, data, kwargs = False, [], {}
            kwargs = {k: v for k, v in args.items() if k in self.params and v}
            page = args['page']
            if kwargs or not page:
                data = self.obj.get(**kwargs)
            else:
                query = []
                per_page = args['per_page']
                if per_page:
                    query = self.obj.get(page=page, per_page=per_page, **kwargs)
                else:
                    query = self.obj.get(page=page, **kwargs)
                if query:
                    data, pages = query[0], query[1]
            return {'totalpage': pages, 'data': data}, 200
        except SakuraAPIError as e:
            app.logger.error(logmsg(traceback.format_exc()))
            return {'error': e.message, 'status': 1}, e.code
        except Exception as e:
            app.logger.error(logmsg(traceback.format_exc()))
            return {'error': str(e), 'status': 1}, 500


class SakuraAPI(Resource):
    """
        Super Task Restful API.
        For both POST(Execute) and GET(Check).
    """
    __abstract__ = True

    def __init__(self, task_name, post_params=None):
        super(SakuraAPI, self).__init__()
        self.parser = reqparse.RequestParser(bundle_errors=True)
        self.task = SakuraTask()
        setattr(self, 'task_name', task_name)
        setattr(self, 'post_params', post_params if post_params else {})

    def _parse_args(self, params):
        for k, v in params.items():
            help = ''
            type_name = v['type'].__name__ if (
                isclass(v['type']) or isfunction(v['type'])) \
                else v['type'].__class__.__name__
            if 'help' in v:
                help = v.pop('help')
            elif 'choices' in v:
                help = '{} must be {} within {}'.format(
                    k, type_name, v['choices'])
            else:
                help = '{} must be {}'.format(k, type_name)
            self.parser.add_argument(name=k, help=help, **v)
        try:
            args = self.parser.parse_args(strict=True)
        except Exception as e:
            if hasattr(e, 'data'):
                raise SakuraInvalidAccessError(e.data['message'])
            else:
                raise SakuraInvalidAccessError()
        args = {k: v for k, v in args.items() if v is not None}
        return args

    def _before_task(self, args):
        return args

    def post(self):
        """ execute a task
        """
        try:
            args = self._parse_args(params=self.post_params)
            args = self._before_task(args=args)
            task = eval('self.task.%s' % self.task_name).apply_async(kwargs=args)
            return {'id': task.task_id, 'status': 0}, 201
        except SakuraAPIError as e:
            app.logger.error(logmsg(traceback.format_exc()))
            return {'error': e.message, 'status': 1}, e.code
        except Exception as e:
            app.logger.error(logmsg(traceback.format_exc()))
            return {'error': str(e), 'status': 1}, 500

    def get(self, id):
        """ check a specific task status
        """
        try:
            task = eval('self.task.%s' % self.task_name).AsyncResult(id)
            result = {
                'id': task.id,
                'state': task.result[0] if task.ready() else task.state,
                'info': task.result[1] if task.ready() else task.result
            }
            return {'result': result, 'status': 0}, 200
        except SakuraAPIError as e:
            app.logger.error(logmsg(traceback.format_exc()))
            return {'error': e.message, 'status': 1}, e.code
        except Exception as e:
            app.logger.error(logmsg(traceback.format_exc()))
            return {'error': str(e), 'status': 1}, 500


class ConfigurationUpdateAPI(SakuraAPI):
    """
        Configuration Update Restful API.
        Inherits from SakuraAPI.
    """
    def __init__(self):
        params = dict(
            service_name=dict(type=str, required=True),
            env_name=dict(type=str, required=True),
            service_version=dict(type=str, required=True),
            check_cmd=dict(type=str),
            reload_cmd=dict(type=str),
            files=dict(
                type=defined_dictionary(
                    keys=dict(
                        name=str, dir=str, mode=str,
                        owner=defined_dictionary(
                            keys=dict(name=str, group=str)),
                        template=unicode,
                        items=union_dictionary(
                            key_type=unicode, value_type=unicode))),
                action='append', required=True),
            hosts=dict(type=ip, action='append', required=True))
        super(ConfigurationUpdateAPI, self).__init__(
            task_name=C.TASK_NAME.CONFIGURATION_UPDATE.value,
            post_params=params)

    def _before_task(self, args):
        tasks = self.task.get(name=self.task_name, ack_status=None)
        tasks.extend(self.task.get(
            name=self.task_name,
            ack_status=C.CONFIGURATION_UPDATE_ACK_STATE.PENDING.value))
        # check pre task
        for x in tasks:
            kwargs = json.loads(x['kwargs'])
            if (kwargs['service_name'] == args['service_name'] and
                    kwargs['env_name'] == args['env_name'] and
                    kwargs['service_version'] == args['service_version']):
                raise SakuraConstraintConflictError(
                    'Pre Task Unconfirmed: {0}'.format(x['task_id']))
        return args


class ConfigurationCheckAPI(SakuraAPI):
    """
        Configuration Check Restful API.
        Inherits from SakuraAPI.
    """
    def __init__(self):
        params = dict(
            files=dict(
                type=defined_dictionary(
                    keys=dict(
                        name=unicode, dir=unicode, mode=str,
                        owner=defined_dictionary(
                            keys=dict(name=str, group=str)),
                        template=unicode,
                        items=union_dictionary(
                            key_type=unicode, value_type=unicode)
                        )
                    ),
                action='append', required=True),
            hosts=dict(type=ip, action='append', required=True))
        super(ConfigurationCheckAPI, self).__init__(
            task_name=C.TASK_NAME.CONFIGURATION_CHECK.value,
            post_params=params)


class ConfigurationAcknowledgeAPI(SakuraAPI):
    """
        Configuration Change Acknowledge Restful API.
        Inherits from SakuraAPI.
    """
    def __init__(self):
        params = dict(main_task_id=dict(type=str, required=True))
        super(ConfigurationAcknowledgeAPI, self).__init__(
            task_name=C.TASK_NAME.CONFIGURATION_ACKNOWLEDGE.value,
            post_params=params)

    def _before_task(self, args):
        # check main task
        main_task = self.task.getObject(task_id=args['main_task_id'])
        if main_task:
            args.pop('main_task_id')
            if main_task.name != C.TASK_NAME.CONFIGURATION_UPDATE:
                raise SakuraInvalidAccessError(
                    'Illegal Main Task Name: {0}.'.format(main_task.name))
            elif main_task.state not in (
                    C.TASK_STATE.SUCCESS.value, C.TASK_STATE.FAILURE.value):
                raise SakuraConstraintConflictError(
                    'Main Task Still Running: {0}.'.format(
                        main_task.task_id))
            elif main_task.ack_status:
                raise SakuraConstraintConflictError(
                    'Main Task Already Confirmed: {0}.'.format(
                        main_task.task_id))
            elif main_task.state == C.TASK_STATE.FAILURE:
                raise SakuraInvalidAccessError(
                    'Failure Main Task Forced to Rollback: {0}.'.format(
                        main_task.task_id))
        else:
            raise SakuraObjectNotFoundError(
                'Main Task Not Found: {0}.'.format(args['main_task_id']))
        # update main task ack_status
        self.task.update(
            task_id=main_task.task_id,
            ack_status=C.CONFIGURATION_UPDATE_ACK_STATE.PENDING.value)
        args['main_task'] = main_task
        return args


class ConfigurationRollbackAPI(SakuraAPI):
    """
        Configuration Rollback Restful API.
        Inherits from SakuraAPI.
    """
    def __init__(self):
        params = dict(main_task_id=dict(type=str, required=True))
        super(ConfigurationRollbackAPI, self).__init__(
            task_name=C.TASK_NAME.CONFIGURATION_ROLLBACK.value,
            post_params=params)

    def _before_task(self, args):
        # check main task
        main_task = self.task.getObject(task_id=args['main_task_id'])
        if main_task:
            args.pop('main_task_id')
            if main_task.name != C.TASK_NAME.CONFIGURATION_UPDATE:
                raise SakuraInvalidAccessError(
                    'Illegal Main Task Name: {0}.'.format(main_task.name))
            elif main_task.state not in (
                    C.TASK_STATE.SUCCESS.value, C.TASK_STATE.FAILURE.value):
                raise SakuraConstraintConflictError(
                    'Main Task Still Running: {0}.'.format(
                        main_task.task_id))
            elif main_task.ack_status:
                raise SakuraConstraintConflictError(
                    'Main Task Already Confirmed: {0}.'.format(
                        main_task.task_id))
        else:
            raise SakuraObjectNotFoundError(
                'Main Task Not Found: {0}.'.format(args['main_task_id']))
        # update main task ack_status
        self.task.update(
            task_id=main_task.task_id,
            ack_status=C.CONFIGURATION_UPDATE_ACK_STATE.PENDING.value)
        args['main_task'] = main_task
        return args
